source/code/plugin/out_mdm.rb

```diff
@@ -57,51 +66,102 @@ def start
-            "client_id" => @data_hash["aadClientId"],
-            "client_secret" => @data_hash["aadClientSecret"],
-            "resource" => @@token_resource_url,
-          }
-        )
-
-        token_response = http_access_token.request(token_request)
-        # Handle the case where the response is not 200
-        parsed_json = JSON.parse(token_response.body)
-        @token_expiry_time = Time.now + 59 * 60 # set the expiry time to be ~one hour from current time
-        @cached_access_token = parsed_json["access_token"]
+      if (Time.now > @get_access_token_backoff_expiry)
+        http_access_token = nil
+        retries = 0
+        begin
+          if @cached_access_token.to_s.empty? || (Time.now + 5 * 60 > @token_expiry_time) # Refresh token 5 minutes from expiration
+            @log.info "Refreshing access token for out_mdm plugin.."
+
+            if (!!@useMsi)
+              @log.info "Using msi to get the token to post MDM data"
+              ApplicationInsightsUtility.sendCustomEvent("AKSCustomMetricsMDMToken-MSI", {})
+              @log.info "Opening TCP connection"
+              http_access_token = Net::HTTP.start(@parsed_token_uri.host, @parsed_token_uri.port, :use_ssl => false)
+              # http_access_token.use_ssl = false
+              token_request = Net::HTTP::Get.new(@parsed_token_uri.request_uri)
+              token_request["Metadata"] = true
+            else
+              @log.info "Using SP to get the token to post MDM data"
+              ApplicationInsightsUtility.sendCustomEvent("AKSCustomMetricsMDMToken-SP", {})
+              @log.info "Opening TCP connection"
+              http_access_token = Net::HTTP.start(@parsed_token_uri.host, @parsed_token_uri.port, :use_ssl => true)
+              # http_access_token.use_ssl = true
+              token_request = Net::HTTP::Post.new(@parsed_token_uri.request_uri)
+              token_request.set_form_data(
+                {
+                  "grant_type" => @@grant_type,
+                  "client_id" => @data_hash["aadClientId"],
+                  "client_secret" => @data_hash["aadClientSecret"],
+                  "resource" => @@token_resource_url,
+                }
+              )
+            end
+
+            @log.info "making request to get token.."
+            token_response = http_access_token.request(token_request)
+            # Handle the case where the response is not 200
+            parsed_json = JSON.parse(token_response.body)
+            @token_expiry_time = Time.now + 29 * 60 # set the expiry time to be ~thirty minutes from current time
+            @cached_access_token = parsed_json["access_token"]
+          @log.info "Successfully got access token"
+          end
+        rescue => err
+          @log.info "Exception in get_access_token: #{err}"
+          if (retries < 2)
+            retries += 1
+            @log.info "Retrying request to get token - retry number: #{retries}"
+            sleep(retries)
```

Comment:
intention here to sleep 1 second, 2 seconds and 3 second before retries?

Reply:
Yes. it is exponential backoff.